src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.cpp

```diff
@@ -108,10 +108,18 @@ Value OnnxToKrnlBuilder::transpose(const Value input,
   return transposeRes;
 }
 
+bool isScalarValue(Value value) {
+  ShapedType stype = value.getType().dyn_cast<ShapedType>();
+  assert(stype && "expected shaped type");
+  return stype.getRank() == 0;
```

Comment:
The onnx spec is unclear about scalar tensor. In case of [ONNXDequantizeLinear](https://github.com/onnx/onnx/blob/main/docs/Operators.md#inputs-2---3-2), it seems 1-D tensor is not scalar:
```
x_scale : tensor(float)
Scale for input 'x'. It can be a scalar, which means a per-tensor/layer dequantization, or a 1-D tensor for per-axis dequantization.
```

Reply:
yes, there are definitely cases where rank 0 scalars represent scalars

[SplitToSequence](https://github.com/onnx/onnx/blob/main/docs/Operators.md#splittosequence) is a rare example where the onnx spec spells out that a scalar should have rank 0

just yesterday I was working on [MatMulInteger](https://github.com/onnx/onnx/blob/main/docs/Operators.md#matmulinteger) and I thought, surely, "scalar" zero points were supposed to be represented by rank 0 tensors because 1-D tensors seemed to represent a distinct notion of "N element vector", but then I discovered that [the example in the spec](https://github.com/onnx/onnx/blob/main/docs/Operators.md#examples-84) expressed scalars a rank 1 tensors

Reply:
It would be great to tighten the definition of scalar to be have only one representation, e.g. always a rank of 0, or a rank of 1 with size 1... (preference for the first option, but its really about the same).

I have a similar beef with similar operations (e.g. Conv vs ConvInt) which use `X `& `W` vs `x` and `w`, which is a pain when you use template (as you cannot reuse a single template if the names are different).

While it's easy to provide "fix" for this, it only moderately help as if we had this new standard for opset 19, we still have to deal with the legacy ops. But maybe it is still better as we can then have a rule that update the older to newer upfront in the frontend, and then only deal with the new names and scalar definitions?

What do you think?

Reply:
it's a shame that the spec was ambiguous an inconsistent to begin with and now we have to deal with legacy ops (and legacy interpretations of the ambiguous spec)

I think we could begin with running through the spec and clarify the spec to make explicit where scalars can be represented in either way, which might involve scanning the examples in the onnx repo and how ops are used in popular models

and do the same for some other ambiguities in the spec

it's not clear to me that we gain much by adding constraints on existing ops in future opsets, for the reasons you mention

Reply:
For ONNX Compiler we can use only one type, say `tensor<f32>` for scalar, and it is the work of the importer to normalize multiple specs of ONNX scalar into the only type in the compiler. 

For tensor of rank 0, e.g. `tensor<0xf32>`, it seems to represent an empty typed tensor in the ONNX specs. So we also have two ways for empty tensor in onnx-mlir: `tensor<0xf32>` and `none`.

Reply:
> For ONNX Compiler we can use only one type, say `tensor<f32>` for scalar, and it is the work of the importer to normalize multiple specs of ONNX scalar into the only type in the compiler.

that's difficult in cases like MatMulInteger zero point arguments which can be model inputs or results of other ops: if we're given a model with shape [1] for such a scalar, we can't easily change the shape to the empty shape (rank 0)

> For tensor of rank 0, e.g. `tensor<0xf32>`, it seems to represent an empty typed tensor in the ONNX specs. So we also have two ways for empty tensor in onnx-mlir: `tensor<0xf32>` and `none`.

actually, tensor<0xf32> has rank 1 and shape [0] (which is "empty" in the sense that getNumElements() == 0)

I propose that we treat a shape with zero elements as any other tensor and that we don't treat it as equivalent to 'none', and reserve 'none' to mean that an input or result is absent - we almost follow that principle everywhere, except in Resize decompositions which I addressed in PR #2133

tensor<f32> has rank 0 with shape [] (and getNumElements() == 1)

Reply:
The disadvantage of `none` is we don't know the element type that we might need during the lowering. I encountered this issue before, but didn't remember how it happened unfortunately.

Reply:
The disadvantage of none is we don't know the element type that we might need during the lowering. I encountered this issue before, but didn't remember how it happened unfortunately.

Reply:
that's interesting, I hadn't thought about that - if you think this is something we should address, please bring it up in the PR #2133 code review